dex/networks/zec/tx.go

```diff
@@ -0,0 +1,611 @@
+func writeTxIn(w io.Writer, version int32, ti *wire.TxIn) error {
+	err := writeOutPoint(w, version, &ti.PreviousOutPoint)
+	if err != nil {
+		return err
+	}
+
+	err = wire.WriteVarBytes(w, pver, ti.SignatureScript)
+	if err != nil {
+		return err
+	}
+
+	return putUint32(w, ti.Sequence)
+}
+
+// writeOutPoint encodes op to the bitcoin protocol encoding for an OutPoint
+// to w.
+func writeOutPoint(w io.Writer, version int32, op *wire.OutPoint) error {
+	_, err := w.Write(op.Hash[:])
+	if err != nil {
+		return err
+	}
+	return putUint32(w, op.Index)
+}
+
+// putUint32 writes a little-endian encoded uint32 to the Writer.
+func putUint32(w io.Writer, v uint32) error {
+	b := make([]byte, 4)
+	binary.LittleEndian.PutUint32(b, v)
+	_, err := w.Write(b)
+	return err
+}
+
+// putUint32 writes a little-endian encoded uint64 to the Writer.
+func putUint64(w io.Writer, v uint64) error {
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, v)
+	_, err := w.Write(b)
+	return err
+}
+
+// readUint32 reads a little-endian encoded uint32 from the Reader.
+func readUint32(r io.Reader) (uint32, error) {
+	b := make([]byte, 4)
+	if _, err := io.ReadFull(r, b); err != nil {
+		return 0, err
+	}
+	return binary.LittleEndian.Uint32(b), nil
+}
+
+// readUint64 reads a little-endian encoded uint64 from the Reader.
+func readUint64(r io.Reader) (uint64, error) {
+	b := make([]byte, 8)
+	if _, err := io.ReadFull(r, b); err != nil {
+		return 0, err
+	}
+	return binary.LittleEndian.Uint64(b), nil
+}
+
+// readTxIn reads the next sequence of bytes from r as a transaction input.
+func readTxIn(r io.Reader, ti *wire.TxIn) error {
+	err := readOutPoint(r, &ti.PreviousOutPoint)
+	if err != nil {
+		return err
+	}
+
+	ti.SignatureScript, err = readScript(r)
+	if err != nil {
+		return err
+	}
+
+	ti.Sequence, err = readUint32(r)
+	return err
+}
+
+// readTxOut reads the next sequence of bytes from r as a transaction output.
+func readTxOut(r io.Reader, to *wire.TxOut) error {
+	v, err := readUint64(r)
+	if err != nil {
+		return err
+	}
+	to.Value = int64(v)
+
+	to.PkScript, err = readScript(r)
+	return err
+}
+
+// readOutPoint reads the next sequence of bytes from r as an OutPoint.
+func readOutPoint(r io.Reader, op *wire.OutPoint) error {
+	_, err := io.ReadFull(r, op.Hash[:])
+	if err != nil {
+		return err
+	}
+
+	op.Index, err = readUint32(r)
+	return err
+}
+
+// readScript reads a variable length byte array. Copy of unexported
+// btcd/wire.readScript.
+func readScript(r io.Reader) ([]byte, error) {
+	count, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return nil, err
+	}
+	if count > uint64(wire.MaxMessagePayload) {
+		return nil, fmt.Errorf("larger than the max allowed size "+
+			"[count %d, max %d]", count, wire.MaxMessagePayload)
+	}
+	b := make([]byte, count)
+	_, err = io.ReadFull(r, b)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func discardBytes(r io.Reader, n int64) error {
+	m, err := io.CopyN(io.Discard, r, n)
```

Comment:
Nice, no need for a buffer or other type, just a reader. io.Discard didn't cross my mind